src/Workspaces/Core/Portable/Storage/CloudCache/CloudCachePersistentStorage.cs

```diff
@@ -0,0 +1,163 @@
+
+        protected override Task<bool> ChecksumMatchesAsync(DocumentKey documentKey, Document? bulkLoadSnapshot, string name, Checksum checksum, CancellationToken cancellationToken)
+            => ChecksumMatchesAsync(name, checksum, GetContainerKey(documentKey, bulkLoadSnapshot), cancellationToken);
+
+        private async Task<bool> ChecksumMatchesAsync(string name, Checksum checksum, CloudCacheContainerKey? containerKey, CancellationToken cancellationToken)
+        {
+            if (containerKey == null)
+                return false;
+
+            using var bytes = s_byteArrayPool.GetPooledObject();
+            checksum.WriteTo(bytes.Object);
+
+            return await _cacheService.CheckExistsAsync(new CloudCacheItemKey(containerKey.Value, name, bytes.Object.AsMemory()), cancellationToken).ConfigureAwait(false);
+        }
+
+        public override Task<Stream?> ReadStreamAsync(string name, Checksum? checksum, CancellationToken cancellationToken)
+            => ReadStreamAsync(name, checksum, s_solutionKey, cancellationToken);
+
+        protected override Task<Stream?> ReadStreamAsync(ProjectKey projectKey, Project? bulkLoadSnapshot, string name, Checksum? checksum, CancellationToken cancellationToken)
+            => ReadStreamAsync(name, checksum, GetContainerKey(projectKey, bulkLoadSnapshot), cancellationToken);
+
+        protected override Task<Stream?> ReadStreamAsync(DocumentKey documentKey, Document? bulkLoadSnapshot, string name, Checksum? checksum, CancellationToken cancellationToken)
+            => ReadStreamAsync(name, checksum, GetContainerKey(documentKey, bulkLoadSnapshot), cancellationToken);
+
+        private async Task<Stream?> ReadStreamAsync(string name, Checksum? checksum, CloudCacheContainerKey? containerKey, CancellationToken cancellationToken)
+        {
+            if (containerKey == null)
+                return _mustSucceed ? throw new InvalidOperationException("Could not create container key") : null;
+
+            if (checksum == null)
+            {
+                return await ReadStreamAsync(new CloudCacheItemKey(containerKey.Value, name), cancellationToken).ConfigureAwait(false);
+            }
+            else
+            {
+                using var bytes = s_byteArrayPool.GetPooledObject();
+                checksum.WriteTo(bytes.Object);
+
+                return await ReadStreamAsync(new CloudCacheItemKey(containerKey.Value, name, bytes.Object.AsMemory()), cancellationToken).ConfigureAwait(false);
+            }
+        }
+
+        private async Task<Stream?> ReadStreamAsync(CloudCacheItemKey key, CancellationToken cancellationToken)
+        {
+            var pipe = new Pipe();
+            var result = await _cacheService.TryGetItemAsync(key, pipe.Writer, cancellationToken).ConfigureAwait(false);
+            if (!result)
+                return null;
+
+            return pipe.Writer.AsStream();
```

Comment:
This is wrong. You should be reading from `pipe.Reader` -- not `pipe.Writer`.

Reply:
whoopsie!  (this is why i need to port our tests to run on this code).   that will happen tomorrow.